### PR TITLE
feat(types): add checksum interface

### PR DESCRIPTION
**Pull request by AndrewFossAWS**
### Description
Add checksum interface

### Testing
yarn test:all

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/types/src/checksum.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+export interface Checksum {
+  /**
+   * Constant length of the digest created by the algorithm in bytes.
+   */
+  digestLength?: string;
+
+  /**
+   * Creates a new checksum object that contains a deep copy of the internal
+   * state of the current `Checksum` object.
+   * This is required by certain languages like Java.
```
Comment (syall) [inline packages/types/src/checksum.ts:23]: Do we need this comment?

```suggestion
```

File: packages/types/src/checksum.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+   */
+  copy?: () => Checksum;
+
+  /**
+   * Returns the digest of all of the data passed.
+   *
+   * This interface may be async, following digest interface standards
+   * for language.
+   * For the purpose of this document, we assume it's called synchronously.
```
Comment (syall) [inline packages/types/src/checksum.ts:32]: Remove this comment.

File: packages/types/src/checksum.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ * The checksum may be performed incrementally as chunks are received or all
+ * at once when the checksum is finalized, depending on the underlying
+ * implementation.
+ *
+ * It's recommended to compute checksum incrementally to avoid reading the
+ * entire payload in memory.
+ *
+ * The constructor may accept an optional secret key while computing checksum
+ * value, when using HMAC. If provided, this secret key would be used when
+ * computing checksum.
```
Comment (syall) [inline packages/types/src/checksum.ts:12]: Maybe document it for classes that implement this interface

```suggestion
 *
 * A class that implements this interface may accept an optional secret key in its
 * constructor while computing checksum value, when using HMAC. If provided,
 * this secret key would be used when computing checksum.
```

**Review by syall — CHANGES_REQUESTED** [review]

**Review by syall — APPROVED** [review]

File: packages/types/src/checksum.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  /**
+   * Constant length of the digest created by the algorithm in bytes.
+   */
+  digestLength?: number;
+
+  /**
+   * Creates a new checksum object that contains a deep copy of the internal
+   * state of the current `Checksum` object.
+   */
+  copy?: () => Checksum;
```
Comment (trivikr) [inline packages/types/src/checksum.ts:24]: Use method syntax
```suggestion
  copy?(): Checksum;
```

Comment (trivikr) [inline packages/types/src/checksum.ts:24]: Use method syntax in interface
```suggestion
  copy?(): Checksum;
```

File: packages/types/src/checksum.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+  /**
+   * Returns the digest of all of the data passed.
+   */
+  digest: () => Promise<Uint8Array>;
```
Comment (trivikr) [inline packages/types/src/checksum.ts:29]: Use method syntax in interface
```suggestion
  digest(): Promise<Uint8Array>;
```

File: packages/types/src/checksum.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  /**
+   * Allows marking a checksum for checksums that support the ability
+   * to mark and reset.
+   *
+   * @param {number} readLimit - The maximum limit of bytes that can be read
+   *   before the mark position becomes invalid.
+   */
+  mark?: (readLimit: number) => void;
```
Comment (trivikr) [inline packages/types/src/checksum.ts:38]: Use method syntax in interface
```suggestion
  mark?(readLimit: number): void;
```

File: packages/types/src/checksum.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+  /**
+   * Resets the checksum to its initial value.
+   */
+  reset: () => void;
```
Comment (trivikr) [inline packages/types/src/checksum.ts:43]: Use method syntax in interface
```suggestion
  reset(): void;
```

**Review by trivikr — CHANGES_REQUESTED** [review]

File: packages/types/src/crypto.ts [hunk 8]
```diff
@@ -4,6 +4,8 @@ export type SourceData = string | ArrayBuffer | ArrayBufferView;
  * An object that provides a hash of data provided in chunks to `update`. The
  * hash may be performed incrementally as chunks are received or all at once
  * when the hash is finalized, depending on the underlying implementation.
+ *
+ * @deprecated use {@link Checksum}
```
Comment (trivikr) [inline packages/types/src/crypto.ts:8]: Add `import { Checksum } from "./checksum";` at the top of the file so that the type would be linked

**Review by trivikr — CHANGES_REQUESTED** [review]

**Review by trivikr — APPROVED** [review]

**Pull request merged**